okhttp/src/commonMain/kotlin/okhttp3/ResponseBody.kt

```diff
@@ -0,0 +1,128 @@
+ * may ultimately cause the application to slow down or crash.
+ *
+ * Both this class and [Response] implement [Closeable]. Closing a response simply
+ * closes its response body. If you invoke [Call.execute] or implement [Callback.onResponse] you
+ * must close this body by calling any of the following methods:
+ *
+ * * `Response.close()`
+ * * `Response.body().close()`
+ * * `Response.body().source().close()`
+ * * `Response.body().charStream().close()`
+ * * `Response.body().byteStream().close()`
+ * * `Response.body().bytes()`
+ * * `Response.body().string()`
+ *
+ * There is no benefit to invoking multiple `close()` methods for the same response body.
+ *
+ * For synchronous calls, the easiest way to make sure a response body is closed is with a `try`
+ * block. With this structure the compiler inserts an implicit `finally` clause that calls
+ * [close()][Response.close] for you.
+ *
+ * ```java
+ * Call call = client.newCall(request);
+ * try (Response response = call.execute()) {
+ * ... // Use the response.
+ * }
+ * ```
+ *
+ * You can use a similar block for asynchronous calls:
+ *
+ * ```java
+ * Call call = client.newCall(request);
+ * call.enqueue(new Callback() {
+ *   public void onResponse(Call call, Response response) throws IOException {
+ *     try (ResponseBody responseBody = response.body()) {
+ *     ... // Use the response.
+ *     }
+ *   }
+ *
+ *   public void onFailure(Call call, IOException e) {
+ *   ... // Handle the failure.
+ *   }
+ * });
+ * ```
+ *
+ * These examples will not work if you're consuming the response body on another thread. In such
+ * cases the consuming thread must call [close] when it has finished reading the response
+ * body.
+ *
+ * ### The response body can be consumed only once.
+ *
+ * This class may be used to stream very large responses. For example, it is possible to use this
+ * class to read a response that is larger than the entire memory allocated to the current process.
+ * It can even stream a response larger than the total storage on the current device, which is a
+ * common requirement for video streaming applications.
+ *
+ * Because this class does not buffer the full response in memory, the application may not
+ * re-read the bytes of the response. Use this one shot to read the entire response into memory with
+ * [bytes] or [string]. Or stream the response with either [source], [byteStream], or [charStream].
+ */
+expect abstract class ResponseBody constructor() : Closeable {
+  abstract fun contentType(): MediaType?
+
+  /**
+   * Returns the number of bytes in that will returned by [bytes], or [byteStream], or -1 if
+   * unknown.
+   */
+  abstract fun contentLength(): Long
+
+  abstract fun source(): BufferedSource
+
+  override fun close()
+
+  companion object {
+    /**
+     * Returns a new response body that transmits this string. If [contentType] is non-null and
+     * lacks a charset, this will use UTF-8.
```

Comment:
There’s a catch here. We won’t honor charsetString on non-JVM platforms because we don’t have an encoder.

Should we crash in such cases?

Reply:
(I think we probably crash on the JVM if we’re given an unknown charset. We should emulate that behavior on other platforms, throwing an exception to indicate that the charset is unknown.)